alica_engine/src/engine/BasicBehaviour.cpp

```diff
@@ -92,168 +56,108 @@ essentials::IdentifierConstPtr BasicBehaviour::getOwnId() const
  */
 bool BasicBehaviour::stop()
 {
-    {
-        std::lock_guard<std::mutex> lck(_runLoopMutex);
-        setSignalState(SignalState::STOP);
-        setStopCalled(true);
-    }
-    if (_behaviour->isEventDriven()) {
-        _runCV.notify_all();
+    if (!isActive(_signalState.load())) {
+        return true;
     }
+    ++_signalState;
+    _engine->editScheduler().schedule(std::bind(&BasicBehaviour::terminateJob, this));
     return true;
 }
 
 /**
  * Starts the execution of this BasicBehaviour.
  */
-bool BasicBehaviour::start()
+bool BasicBehaviour::start(RunningPlan* rp)
 {
-    {
-        std::lock_guard<std::mutex> lck(_runLoopMutex);
-        setSignalState(SignalState::START);
-    }
-
-    if (!_behaviour->isEventDriven()) {
-        _runCV.notify_all();
+    if (isActive(_signalState.load())) {
+        return true;
     }
+    ++_signalState;
+    // This has to be done after incrementing _signalState for correct behaviour of getPlanContext()
```

Comment:
Better to remove this comment since it is not entirely accurate. See getPlanContext()'s comment in the header file